### PR TITLE
sources/azure: address mypy/pyright typing complaints

### PR DESCRIPTION
**Pull request by cjp256**

### CODE REVIEW

**Review by anhvoms — APPROVED** [review]

**Comment by cjp256** [comment]
Depends on #1232, #1222

**Review by anhvoms — APPROVED** [review]

File: cloudinit/sources/DataSourceAzure.py [hunk 1]
```diff
@@ -2292,6 +2309,8 @@ def _get_metadata_from_imds(
             exception_cb=exc_cb,
             infinite=infinite,
         )
+        if response is None:
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceAzure.py:2312]: Same comment about readurl

File: cloudinit/sources/DataSourceAzure.py [hunk 2]
```diff
@@ -2203,7 +2217,7 @@ def _generate_network_config_from_imds_metadata(imds_metadata) -> dict:
                 privateIp = addr["privateIpAddress"]
                 if not dev_config.get("addresses"):
                     dev_config["addresses"] = []
-                dev_config["addresses"].append(
+                dev_config["addresses"].append(  # type: ignore
```
Comment (TheRealFalcon) [inline cloudinit/sources/DataSourceAzure.py:2220]: Hmm, using pylance these 3 in this section don't trigger anything. Curious what mypy doesn't like.

**Review by TheRealFalcon — COMMENTED** [review]
I left a number of comments inline. Take or leave what you'd like.

File: cloudinit/distros/__init__.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 import string
 import urllib.parse
 from io import StringIO
-from typing import Any, Mapping  # noqa: F401
+from typing import Any, Mapping, Type  # noqa: F401
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/distros/__init__.py]: good call, fixed!

File: cloudinit/sources/DataSourceAzure.py [hunk 4]
```diff
@@ -547,9 +549,9 @@ def crawl_metadata(self):
         if metadata_source == "IMDS" and not crawled_data["files"]:
             try:
                 contents = build_minimal_ovf(
-                    username=imds_username,
-                    hostname=imds_hostname,
-                    disableSshPwd=imds_disable_password,
+                    username=imds_username,  # type: ignore
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py:552]: I planned to revisit this one with another small refactor.  I wanted to spend a few more cycles making sure there wasn't an edge case where "None" was making it into the generated OVF harmlessly before affecting control flow.  I'll come back to this one.

File: cloudinit/sources/helpers/azure.py [hunk 5]
```diff
@@ -382,6 +382,8 @@ def http_with_retries(url, **kwargs) -> str:
                 )
             time.sleep(sleep_duration_between_retries)
 
+    if exc is None:
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/sources/helpers/azure.py]: Agreed, fixed!

File: cloudinit/sources/DataSourceAzure.py [hunk 6]
```diff
@@ -1304,14 +1317,17 @@ def exc_cb(msg, exception):
...
+                    response = readurl(
                         url,
                         timeout=IMDS_TIMEOUT_IN_SECONDS,
                         headers=headers,
                         exception_cb=exc_cb,
                         infinite=True,
                         log_req_resp=False,
-                    ).contents
+                    )
+                    if response is None:
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py]: Agreed, fixed! Lmk if this is better :D

File: cloudinit/sources/DataSourceAzure.py [hunk 7]
```diff
@@ -1921,7 +1937,7 @@ def read_azure_ovf(contents):
         elif name == "userpassword":
             password = value
         elif name == "hostname":
-            md["local-hostname"] = value
+            md["local-hostname"] = value  # type: ignore
```
[the remark this answers is not in the document]
Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py]: Great point, thanks! Fixed

File: cloudinit/sources/DataSourceAzure.py (hunk 1 above)
Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py]: Fixed!

File: cloudinit/sources/DataSourceAzure.py (hunk 2 above)
Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py:2217]: cloudinit/sources/DataSourceAzure.py:2217: error: "object" has no attribute "append"

Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py:2217]: pyright makes no complaints about it either

Comment (cjp256) [reply cloudinit/sources/DataSourceAzure.py]: Setting the dictionaries to Dict[str, Any] seems to do the trick as well for these :D

File: cloudinit/sources/helpers/azure.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (TheRealFalcon) [reply cloudinit/sources/helpers/azure.py]: We still need to silence the warning in this function. The raise on the last line complains about possibly being unbound.

**Pull request merged**